### PR TITLE
Remove serde support

### PR DESCRIPTION
**Pull request by rozbb**
This PR removes serde support. This is following a [conversation](https://github.com/RustCrypto/utils/pull/979#issuecomment-1788867202) elsewhere about providing serde impls for byte slices. The argument for removing serde impls is:
1. If you derive the plain `[T; N]` serialization, or `(T, T, ..., T)` serialization, then you miss out on optimizations for formats with byteslice representations. This has a horrible slowdown, which I've experience with `wasm_bindgen`. This is why `serde_bytes` exists, and it cannot be merged into upstream (see above conversation)
2. The choice of whether or not to use a length tag on the serialization depends on the format you're serializing to (see above conversation)
3. ~~The current p256 `serde::Serialize` implementation [is the DER encoding](https://docs.rs/elliptic-curve/0.13.6/elliptic_curve/struct.PublicKey.html) of the public key. This is not something I noticed, and certainly doesn't match the raw byte representation of the other public key serializations. I should not have autoderived without looking.~~ I don't think this is true. It's the bytes given by `to_bytes()`.

In short, people have different needs when it comes to serializing HPKE values with `serde`, and any choice necessarily excludes other options. So rather than making that choice for the user, this crate should instead just expose its normal [`Serializable`](https://github.com/rozbb/rust-hpke/blob/57fce26b436f47846ee4f9a972ea0675786101c9/src/lib.rs#L191-L215) trait (with methods `to_bytes` and `write_exact`), and let the user implement what they need.

Looking for feedback on this, since it's removing a feature that might be in use currently.

cc @marmeladema @bwesterb @tarcieri

### CODE REVIEW

**Comment by tarcieri** [comment]
The real kicker is this: https://github.com/serde-rs/serde/issues/2120#issuecomment-1078435661

The `serde` data model lacks the concept of a homogeneously-typed fixed-size array/tuple. This means there is no universally agreed upon compact representation of a fixed-size byte array. The best you can do is `serde::bytes::Bytes` which calls `serialize_bytes` on the underlying serializer, *but* this will add a length prefix even if the length is known in advance.

The only way to avoid a length prefix is to use `serialize_tuple`, however for some message formats like `MessagePack` this will include type information with each element, leading to a `[u8; N]` serialization which is much larger than `N` bytes.

**Comment by rozbb** [comment]
Ya exactly. I notice you said in the linked thread you were switching to `serialize_bytes`. In which crate is that happening? I don't think it's necessarily a good option here, but I'm curious to see.

**Comment by tarcieri** [comment]
It was in the `serdect` crate here: https://github.com/RustCrypto/formats/pull/1112

It's designed specifically for constant-time bytes handling. Using `serde_bytes` was the only way to ensure encoding in MessagePack isn't data-dependent. The additional and often unnecessary length prefix is unfortunate.

**Comment by rozbb** [comment]
It looks like there is one project that is using this, specifically the the DER-encoded p256, [here](https://github.com/khonsulabs/bonsaidb/blob/b110589b9d0f2bc0fcbd60690b1d26b73f41d373/crates/bonsaidb-local/src/vault.rs#L616-L622) and [here](https://github.com/khonsulabs/bonsaidb/blob/b110589b9d0f2bc0fcbd60690b1d26b73f41d373/crates/bonsaidb-local/src/vault.rs#L76-L85).

I think the proper thing to do in this case is to define some helper functions  to deserialize at least p256, put them in `legacy.rs` or something, and mark them for deprecation in the next version. Does that sound appropriate?

**Comment by rozbb** [comment]
Oh, maybe that shouldn't be in here. Because those helpers will be specific to the file format. There needs to be a `bincode_p256_to_canonical()` function.

Ok, so just delete the impls here, and document / offer a workaround?
